### PR TITLE
Material Reclaimer

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

Adds the material reclaimer: a machine capable of converting the junk you have lying around into raw materials.
Simply insert the stuff, give it a second to chug, and watch materials and chemicals pop out.

chemicals can acquired by clicking on the reclaimer with a container of some kind (akin to a fuel tank)

does a few extra things with pricing, too:
- materialPrototype cost now overrides StackPriceComponent
- fixed an oversight in NoMaterialArbitrage that underestimated material cost
- fixed a few prototypes that had too low prices.

TODO
- [x] tests for the hellscape that is PhysicalCompositionComponent

- [x] giving like 75% of entities some kind of composition.

**Media**

https://user-images.githubusercontent.com/98561806/228993935-a048fcff-38b1-4cd1-8061-be875d81effc.mp4

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Changelog**

:cl:
- add: Added the material reclaimer. Insert junk you don't want to grind it up into its raw material form. Try stuff out!
- fix: Fixed a few material exploit methods.
- tweak: Slightly nerfed lathe material reduction upgrades.

### CODE REVIEW

**Comment by Elijahrane** [comment]
isn't this what the recycler is supposed to do

**Comment by EmoGarbage404** [comment]
> isn't this what the recycler is supposed to do

I imagine the material reclaimer as a low-throughput system that has a high return yield.
The recycler can be quicker, but have a more limited yield in return.

I don't plan on making the changes to the recycler in this pr, but I've modified this system to allow for it to be adapted for the recycler (as i envisioned it) in the future.

**Comment by EmoGarbage404** [comment]
Also, open letter to all, does this need a guide entry? If so, what general category should it be listed under?

**Comment by deltanedas** [comment]
i think its fairly self explanatory, may as well make a protolathe guide at that point (dont)

**Comment by EmoGarbage404** [comment]
> i think its fairly self explanatory, may as well make a protolathe guide at that point (dont)

funnily enough, the protolathe has a small guide in the science section.

**Comment by EmoGarbage404** [comment]
Ready for review

**Comment by mirrorcult** [comment]
i think im still kind of confused on why this and the recycler are separate systems, i think that will end up being confusing

**Comment by EmoGarbage404** [comment]
> i think im still kind of confused on why this and the recycler are separate systems, i think that will end up being confusing

I felt that it was important to give players options. In situations where material balance is important and you need to scavenge, you may feel incentivized to carefully shift through trash to find items that give a decent return. Otherwise, if you simply need to destroy trash, you can simply funnel it through a recycler to get rid of it quicker for a much smaller return (or at least that's my intent).

I do plan on merging the two systems; i just wanted to split up the pr since it got big fairly quickly.

I think giving players the option off efficiency vs. returns is enough to justify two similar machines. This one also collects chemicals from reclaimed objects, which can be easily collected.

**Comment by EmoGarbage404** [comment]
Converted recycling to use the same system as the reclaimer. 
Hopefully the relationship between the two machines is more clear now.

File: Content.Client/Materials/MaterialReclaimerSystem.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿using Content.Shared.Materials;
+
+namespace Content.Client.Materials;
+
+public sealed class MaterialReclaimerSystem : SharedMaterialReclaimerSystem
```
Comment (moonheart08) [inline Content.Client/Materials/MaterialReclaimerSystem.cs:5]:
```suggestion
/// <inheritdoc/>
public sealed class MaterialReclaimerSystem : SharedMaterialReclaimerSystem
```

File: Content.Server/Materials/MaterialReclaimerSystem.cs [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+using Content.Shared.Interaction;
+using Content.Shared.Interaction.Events;
+using Content.Shared.Materials;
+using Robust.Server.GameObjects;
+using Robust.Shared.Player;
+using Robust.Shared.Utility;
+
+namespace Content.Server.Materials;
+
+public sealed class MaterialReclaimerSystem : SharedMaterialReclaimerSystem
```
Comment (moonheart08) [inline Content.Server/Materials/MaterialReclaimerSystem.cs:26]:
```suggestion
/// <inheritdoc/>
public sealed class MaterialReclaimerSystem : SharedMaterialReclaimerSystem
```

File: Content.Server/Materials/MaterialReclaimerSystem.cs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        _appearance.SetData(uid, RecyclerVisuals.Bloody, true);
+    }
+
+    private void OnActivePowerChanged(EntityUid uid, ActiveMaterialReclaimerComponent component, ref PowerChangedEvent args)
+    {
+        if (!args.Powered)
+            TryFinishProcessItem(uid, null, component);
+    }
+
+    public override bool TryFinishProcessItem(EntityUid uid, MaterialReclaimerComponent? component = null, ActiveMaterialReclaimerComponent? active = null)
```
Comment (moonheart08) [inline Content.Server/Materials/MaterialReclaimerSystem.cs:129]:
```suggestion
    /// <inheritdoc/>
    public override bool TryFinishProcessItem(EntityUid uid, MaterialReclaimerComponent? component = null, ActiveMaterialReclaimerComponent? active = null)
```

File: Content.Server/Materials/MaterialReclaimerSystem.cs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+        // scales the output if the process was interrupted.
+        var completion = 1f - Math.Clamp((float) Math.Round((active.EndTime - Timing.CurTime) / active.Duration),
+            0f, 1f);
+        Reclaim(uid, item, completion, component);
+
+        return true;
+    }
+
+    public override void Reclaim(EntityUid uid,
```
Comment (moonheart08) [inline Content.Server/Materials/MaterialReclaimerSystem.cs:151]:
```suggestion
    /// <inheritdoc/>
    public override void Reclaim(EntityUid uid,
```

File: Content.Shared/Materials/ActiveMaterialReclaimerComponent.cs [hunk 5]
```diff
@@ -0,0 +1,18 @@
+﻿using Robust.Shared.Containers;
+using Robust.Shared.GameStates;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom;
+
+namespace Content.Shared.Materials;
+
+[RegisterComponent, NetworkedComponent, Access(typeof(SharedMaterialReclaimerSystem))]
+public sealed class ActiveMaterialReclaimerComponent : Component
```
Comment (moonheart08) [inline Content.Shared/Materials/ActiveMaterialReclaimerComponent.cs:8]: where doc comments

File: Content.Shared/Materials/MaterialReclaimerComponent.cs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+using Robust.Shared.GameStates;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype;
+
+namespace Content.Shared.Materials;
+
+[RegisterComponent, NetworkedComponent, Access(typeof(SharedMaterialReclaimerSystem))]
+public sealed class MaterialReclaimerComponent : Component
```
Comment (moonheart08) [inline Content.Shared/Materials/MaterialReclaimerComponent.cs:14]: so this does what? (doc comments)

File: Content.Shared/Materials/SharedMaterialReclaimerSystem.cs [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+using Content.Shared.Mobs.Components;
+using Content.Shared.Stacks;
+using Robust.Shared.Containers;
+using Robust.Shared.GameStates;
+using Robust.Shared.Physics.Events;
+using Robust.Shared.Timing;
+
+namespace Content.Shared.Materials;
+
+public abstract class SharedMaterialReclaimerSystem : EntitySystem
```
Comment (moonheart08) [inline Content.Shared/Materials/SharedMaterialReclaimerSystem.cs:18]: doc comments

File: Content.Shared/Materials/SharedMaterialReclaimerSystem.cs [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    {
+        component.ReclaimingContainer = _container.EnsureContainer<Container>(uid, ActiveReclaimerContainerId);
+    }
+
+    private void OnActiveUnpaused(EntityUid uid, ActiveMaterialReclaimerComponent component, ref EntityUnpausedEvent args)
+    {
+        component.EndTime += args.PausedTime;
+    }
+
+    public bool TryStartProcessItem(EntityUid uid, EntityUid item, MaterialReclaimerComponent? component = null, EntityUid? user = null)
```
Comment (moonheart08) [inline Content.Shared/Materials/SharedMaterialReclaimerSystem.cs:93]: dock comment

File: Content.Shared/Materials/SharedMaterialReclaimerSystem.cs [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+        }
+
+        var active = EnsureComp<ActiveMaterialReclaimerComponent>(uid);
+        active.Duration = duration;
+        active.EndTime = Timing.CurTime + duration;
+        active.ReclaimingContainer.Insert(item);
+        return true;
+    }
+
+    public virtual bool TryFinishProcessItem(EntityUid uid, MaterialReclaimerComponent? component = null, ActiveMaterialReclaimerComponent? active = null)
```
Comment (moonheart08) [inline Content.Shared/Materials/SharedMaterialReclaimerSystem.cs:133]: dooc commemntwert

File: Content.Shared/Materials/SharedMaterialReclaimerSystem.cs [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+    {
+        if (!Resolve(uid, ref component, ref active, false))
+            return false;
+
+        RemCompDeferred(uid, active);
+        return true;
+    }
+
+    public virtual void Reclaim(EntityUid uid,
```
Comment (moonheart08) [inline Content.Shared/Materials/SharedMaterialReclaimerSystem.cs:142]: doc comment

File: Content.Shared/Materials/SharedMaterialReclaimerSystem.cs [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+            return;
+
+        component.ItemsProcessed++;
+        if (component.CutOffSound)
+            component.Stream?.Stop();
+
+        Dirty(component);
+    }
+
+    public void SetReclaimerEnabled(EntityUid uid, bool enabled, MaterialReclaimerComponent? component = null)
```
Comment (moonheart08) [inline Content.Shared/Materials/SharedMaterialReclaimerSystem.cs:157]: d o c comment

File: Content.Shared/Materials/SharedMaterialReclaimerSystem.cs [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+    {
+        if (!Resolve(uid, ref component, false))
+            return;
+        component.Enabled = enabled;
+        AmbientSound.SetAmbience(uid, enabled && component.Powered);
+        Dirty(component);
+    }
+
+    public bool CanStart(EntityUid uid, MaterialReclaimerComponent component)
```
Comment (moonheart08) [inline Content.Shared/Materials/SharedMaterialReclaimerSystem.cs:166]: a smidge of doc comment

File: Content.Shared/Materials/SharedMaterialReclaimerSystem.cs [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+    {
+        if (HasComp<ActiveMaterialReclaimerComponent>(uid))
+            return false;
+
+        return component.Powered && component.Enabled;
+    }
+
+    public bool CanGib(EntityUid uid, EntityUid victim, MaterialReclaimerComponent component)
```
Comment (moonheart08) [inline Content.Shared/Materials/SharedMaterialReclaimerSystem.cs:174]: dock comment

File: Content.Shared/Materials/SharedMaterialReclaimerSystem.cs [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+
+        var materialSum = compositionComponent.MaterialComposition.Values.Sum();
+        materialSum *= CompOrNull<StackComponent>(item)?.Count ?? 1;
+        var duration = TimeSpan.FromSeconds(materialSum / reclaimerComponent.MaterialProcessRate);
+        if (duration < reclaimerComponent.MinimumProcessDuration)
+            duration = reclaimerComponent.MinimumProcessDuration;
+        return duration;
+    }
+
+    public override void Update(float frameTime)
```
Comment (moonheart08) [inline Content.Shared/Materials/SharedMaterialReclaimerSystem.cs:212]:
```suggestion
    /// <inheritdoc/>
    public override void Update(float frameTime)
```

**Review by moonheart08 — CHANGES_REQUESTED** [review]

**Review by moonheart08 — APPROVED** [review]

**Pull request merged**